File: native-benchmarks/native-benchmarks/file/FileOperationsBenchmark.swift
//
//  FileOperationsBenchmark.swift
//  native-benchmarks
//
//  Created by Vanja Vidmark on 2025-04-15.
//

import Foundation

class FileOperationsBenchmark {
    private let fileManager = FileManager.default
    private let performanceCalculator: HardwarePerformanceCalculator
    private let data: Data
    
    init(performanceCalculator: HardwarePerformanceCalculator) {
        self.performanceCalculator = performanceCalculator
        let sizeInMB = 100
        self.data = Data(count: sizeInMB * 1024 * 1024)
    }

    func fileURL(for index: Int, suffix: String? = nil) -> URL {
        let dir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let filename = suffix != nil
            ? "file_\(index)_\(suffix!).dat"
            : "file_\(index).dat"
        return dir.appendingPathComponent(filename)
    }

    func write(index: Int, data: Data, suffix: String? = nil) {
        let url = fileURL(for: index, suffix: suffix)

        do {
            FileManager.default.createFile(atPath: url.path, contents: nil, attributes: nil)
            let fileHandle = try FileHandle(forWritingTo: url)
            defer { try? fileHandle.close()}
            try fileHandle.write(contentsOf: data)
            try fileHandle.synchronize()  // flushes write buffers
        } catch {
            print("Write error at \(url.path): \(error)")
        }
    }

    
    func read(index: Int, suffix: String? = nil) -> Data? {
        let url = fileURL(for: index, suffix: suffix)
        guard fileManager.fileExists(atPath: url.path) else {
            print("File not found: \(url.path)")
            return nil
        }
        
        do {
            let fileHandle = try FileHandle(forReadingFrom: url)
            defer { try? fileHandle.close() }
            return try fileHandle.readToEnd()
        } catch {
            print("Read error at \(url.path): \(error)")
            return nil
        }
    }

    func delete(index: Int, suffix: String? = nil) {
        let url = fileURL(for: index, suffix: suffix)
        if !fileManager.fileExists(atPath: url.path) {
            print("File not found: \(url.path)")
        }
        try? fileManager.removeItem(at: url)
    }

    func runWriteBenchmark(warmup: Int, n: Int, measureTime: Bool) {
        for i in 0..<warmup {
            self.write(index: i, data: data, suffix: "write")
            print("Warmup: Wrote file \(i)")
        }
        if measureTime {
            for i in warmup..<n+warmup {
                let start = ProcessInfo.processInfo.systemUptime
                self.write(index: i, data: data, suffix: "write")
                let duration = ProcessInfo.processInfo.systemUptime - start
                performanceCalculator.sampleTime(duration: duration)
                print("Measured time: Wrote file \(i)")
            }
            performanceCalculator.postTimes()
        } else {
            performanceCalculator.start()
            for i in warmup..<n+warmup {
                performanceCalculator.markIteration(i-warmup)
                self.write(index: i, data: data, suffix: "write")
                print("Measured performance: Wrote file \(i)")
            }
            performanceCalculator.stopAndPost()
        }
        for i in 0..<n+warmup {
            self.delete(index: i, suffix: "write")
            print("Deleted file \(i)")
        }
        print("File write benchmark done")
    }
    
    // Writes all files to be read during the read benchmark
    func preWriteFiles(files: Int) {
        for i in 0..<files {
            self.write(index: i, data: data, suffix: "read")
            print("Wrote file \(i), read")
        }
        print("Files pre-written, restart app!")
    }
    
    func runReadBenchmark(warmup: Int, n: Int, measureTime: Bool) {
        let totalFiles = warmup + n
        var indices = Array(0..<totalFiles)
        indices.shuffle()

        // Warmup reads (not measured)
        for i in 0..<warmup {
            let idx = indices[i]
            autoreleasepool {
                _ = read(index: idx, suffix: "read")
            }
        }
        if measureTime {
            for i in warmup..<n+warmup {
                let start = ProcessInfo.processInfo.systemUptime
                autoreleasepool {
                    _ = read(index: indices[i], suffix: "read")
                }
                let duration = ProcessInfo.processInfo.systemUptime - start
                performanceCalculator.sampleTime(duration: duration)
            }
            performanceCalculator.postTimes()
        } else {
            performanceCalculator.start()
            for i in warmup..<n+warmup {
                performanceCalculator.markIteration(i-warmup)
                autoreleasepool {
                    _ = read(index: indices[i], suffix: "read")
                }
            }
            performanceCalculator.stopAndPost()
        }
        print("File read benchmark done")
    }
}

File: kmp-benchmarks/composeApp/src/commonMain/kotlin/org/example/kmpbenchmarks/file/FileOperationsBenchmark.kt
package org.example.kmpbenchmarks.file

import org.example.kmpbenchmarks.PerformanceCalculator
import org.example.kmpbenchmarks.getTime

expect fun write(index: Int, data: ByteArray, suffix: String? = null)
expect fun delete(index: Int, suffix: String? = null)
expect fun read(index: Int, suffix: String? = null) : ByteArray?

class FileOperationsBenchmark(private val performanceCalculator: PerformanceCalculator) {
    private val sizeInMB = 100
    private val data: ByteArray = ByteArray(sizeInMB * 1024 * 1024)

    fun runWriteBenchmark(warmup: Int, n: Int, measureTime: Boolean) {
        for (i in 0 until warmup) {
            write(i, data, suffix = "write")
            print("Warmup: Wrote file $i \n")
        }

        if (measureTime) {
            for (i in warmup until n+warmup) {
                val start = getTime()
                write(i, data, suffix = "write")
                val duration = getTime() - start
                performanceCalculator.sampleTime(duration)
                print("Measured time: Wrote file $i \n")
            }
            performanceCalculator.postTimes()
        } else {
            performanceCalculator.start()
            for (i in warmup until n+warmup) {
                performanceCalculator.markIteration(i-warmup)
                write(i, data, suffix = "write")
                print("Measured performance: Wrote file $i \n")
            }
            performanceCalculator.stopAndPost()
        }

        for (i in 0 until n+warmup) {
            delete(i, suffix = "write")
            println("Deleted file $i \n")
        }
        println("File write benchmark done")
    }

    fun preWriteFiles(files: Int) {
        for (i in 0 until files) {
            write(i, data, suffix = "read")
            println("Wrote file $i, read")
        }
        println("Files pre-written, restart app!")
    }

    fun runReadBenchmark(warmup: Int, n: Int, measureTime: Boolean) {
        val totalFiles = warmup + n
        val indices = (0 until totalFiles).shuffled()

        // Warmup (not measured)
        for (i in 0 until warmup) {
            val idx = indices[i]
            read(idx, suffix = "read")
        }

        if (measureTime) {
            for (i in warmup until n+warmup) {
                val start = getTime()
                read(indices[i], suffix = "read")
                val duration = getTime() - start
                performanceCalculator.sampleTime(duration)
            }
            performanceCalculator.postTimes()
        } else {
            performanceCalculator.start()
            for (i in warmup until n+warmup) {
                performanceCalculator.markIteration(i-warmup)
                read(indices[i], suffix = "read")
            }
            performanceCalculator.stopAndPost()
        }
        println("File read benchmark done")
    }

}
